### PR TITLE
Integrate sorald buildbreaker as a workflow job

### LINKED ISSUE #50: Integrate sorald with diffmin
I was thinking if we should integrate `sorald` in our workflow. It can tell us about the warnings in our code and can even repair them. One `sorald-buildbreaker` is released, we could integrate that action too.

### PR DESCRIPTION
**Pull request by algomaster99**
Fixes #50

### CODE REVIEW

**Comment by algomaster99** [comment]
```sh
Failed to resolve action download info. Error: Unable to resolve action `SpoonLabs/sorald-buildbreaker@45bafb6`, the provided ref `45bafb6` is the shortened version of a commit SHA, which is not supported. Please use the full commit SHA `45bafb6db29a466080328864cecf9b20d9ace3de` instead.
```

I mean if the action is aware which commit I am referring to, why not just execute it 🤦‍♂️

**Comment by algomaster99** [comment]
@slarse if the sorald has nothing to repair or it repairs everything successfully, will those changed be committed in the PR?

Currently, it is breaking at one place (violating 1118). I will fix that and push the changes.

File: src/main/java/com/diffmin/GumtreeSpoonAstDiff.java [hunk 1]
```diff
@@ -16,6 +16,12 @@
  * taken from.
  */
 public class GumtreeSpoonAstDiff {
+
+    /** Override constructor to prevent instantiating of this class (RSPEC-1118). */
+    private GumtreeSpoonAstDiff() {
```
Comment (algomaster99) [inline src/main/java/com/diffmin/GumtreeSpoonAstDiff.java:21]: Although this constructor will decrease the coverage, should we care writing a test for it? This class has to be made as a library so we can just reuse it.

Comment (slarse) [reply src/main/java/com/diffmin/GumtreeSpoonAstDiff.java:21]: No need to write a test for a private constructor. In fact, that's inconveniently hard (you'd need to change the visibility with reflection or something like that).

**Review by slarse — APPROVED** [review]
> @slarse if sorald repairs everything successfully, will those changes be committed in the PR?

No, right now `sorald-buildbreaker` only breaks the build when it finds something that Sorald can repair. The next level would be to also provide the fixes, but I'm not quite sure how we will do that yet. Just pushing to a PR without asking is kind of rude :)

> Failed to resolve action download info. Error: Unable to resolve action `SpoonLabs/sorald-buildbreaker@45bafb6`, the provided ref `45bafb6` is the shortened version of a commit SHA, which is not supported. Please use the full commit SHA `45bafb6db29a466080328864cecf9b20d9ace3de` instead.

This is probably just a safety mechanism. A short hash is significantly more likely to cause a hash collision than the full hash is.

But yeah, this LGTM. Want to merge?

**Comment by algomaster99** [comment]
> Just pushing to a PR without asking is kind of rude

Agreed!

> A short hash is significantly more likely to cause a hash collision than the full hash is.

I get that but GitHub and git also consider 7 characters to uniquely identify a hash so that's why I was skeptical about the error. 😅 

Yes, we can merge!

**Comment by slarse** [comment]
> > A short hash is significantly more likely to cause a hash collision than the full hash is.
> 
> I get that but GitHub and git also consider 7 characters to uniquely identify a hash so that's why I was skeptical about the error. sweat_smile

I wouldn't put it like that. The short hash is typically considered to be unique enough that the chance for an accidental collision within a project is pretty unlikely. However, the "old" SHA1 was broken by researches (Git now uses a modified SHA1) even at the full 160 bit size (20 characters). But as SHA1 has been broken before, it may be again, it's generally considered to be "not the best" hashing algorithm. There is in fact talks of moving Git to a [stronger hash](https://git-scm.com/docs/hash-function-transition/2.18.0).

Reducing the hash size to just 56 bits obviously makes SHA1 even weaker, and so a malicious agent purposefully engineering a commit with a hash collision becomes more plausible. This is especially dangerous with something like GHA, where the actions potentially gain access to secrets in other projects.

**Comment by slarse** [comment]
@monperrus Thanks to @algomaster99's initiative, we now have sorald-buildbreaker running on diffmin :)

**Comment by algomaster99** [comment]
> Reducing the hash size to just 56 bits obviously makes SHA1 even weaker, and so a malicious agent purposefully engineering a commit with a hash collision becomes more plausible. This is especially dangerous with something like GHA, where the actions potentially gain access to secrets in other projects.

I did not imagine the consequence could be so profound! It is great to learn stuff like this. :)

**Comment by monperrus** [comment]
> Thanks to @algomaster99's initiative, we now have sorald-buildbreaker running on diffmin :)

Oh super nice! Our goal is to have sorald-buildbreaker on all active Java projects in the group (Spoon, Depclean, Pankti, Repairnator, etc.).

**Comment by algomaster99** [comment]
@monperrus Yes, that would be great. It will help in testing `sorald-buildbreaker`.

By the way, what is the difference between [repairnator](https://github.com/eclipse/repairnator) and [sorald](https://github.com/SpoonLabs/sorald)? Sorald seems more focused to me in terms of what repairs it does.

**Comment by slarse** [comment]
@algomaster99 Sorald is an automated program repair tool (APR). Repairnator is more of a framework for applying APR tools, and it makes use of Sorald, among other things.

**Comment by algomaster99** [comment]
Oh, okay. I understand.

**Pull request merged**